docs/architecture.md

```diff
@@ -4,213 +4,217 @@
+| aggregate-store | DynamoDB | `piece`                  | ---        |
+| inclusion-store | DynamoDB | `aggregate`              | ---        |
+
+### Queries and associayed stores
+
+1. Check if piece is already in the pipeline (`inclusion-store` with fallback for `piece-store`)
+  - Before getting a piece into the pipelined queue, it should be stored to guarantee uniqueness
+  - Also important to be able to report back on the "in progress" work when deal state of a piece is still unknown by fallbacking to Head Request in bucket
+2. Put and Get buffer blocks (`buffer-store`)
+  - While `buffer-queue` is concatenating buffers, these blocks will be stored to be propagated through queue stages via their CIDs
+  - Key `${blockCid}/{blockCid}`, Value empty with expiration date
+3. Put aggregates offered to `spade` and get aggregates pending resolution (`aggregate-store`)
+4. Write pieces together with the aggregate they are part of (`inclusion-store`)
+5. Get deal state of a given piece (`inclusion-store` with fallback for `piece-store`)
+  - Rely on DynamoDB to get `aggregate` where the piece is (secondary index) and use it ask Spade for details for the deal
+  - In case there is no record in Dynamo, we can fallback to check S3 bucket, in order to reply that piece is being aggregated if it is there
+
+TODO: metrics
+
+### `piece-store` schema
+
+```typescript
+interface piece {
+  // CID of the piece `bagy...content` (primary index, partition key)
+  piece: PieceCID
+  // number of milliseconds elapsed since the epoch for piece inserted
+  insertedAt: number
+  // identifier of producer to derive `group-id` `did:web:web3.storage` (primary index, sort key)
+  space: string
+  // identifier of group to derive `group-id` `did:web:free.web3.storage`
+  group: string
+}
+```
+
+### `buffer-store` schema
+
+```typescript
+interface PieceBuffer {
+  // Pieces inside the buffer
+  pieces: Piece[]
+}
+
+interface Piece {
+  piece: PieceCID
+  // number of milliseconds elapsed since the epoch when piece was received
+  insertedAt: number
+  // Policies that this piece is under
+  policy: PiecePolicy
+}
+
+type PiecePolicy =
+  | NORMAL
+  | RETRY
+
+type NORMAL = 0
+type RETRY = 1
+```
+
+### `aggregate-store` schema
+
+```typescript
+interface Aggregate {
+  // PieceCid of an Aggregate `bagy...aggregate` (primary index, partition key)
+  piece: PieceCID
+  // CID of dag-cbor block `bafy...cbor`
+  buffer: Link
+  // CID of `aggregate/add` invocation `bafy...inv`
+  invocation: Link
+  // CID of `aggregate/add` task `bafy...task`
+  task: Link
+  // number of milliseconds elapsed since the epoch when aggregate was submitted
+  insertedAt: number
+  // known status of the aggregate (a secondary index)
+  stat: AggregateStatus
+}
+
+type AggregateStatus =
+  | OFFERED
+  | APPROVED
+  | REJECTED
+
+type OFFERED = 0
+type APPROVED = 1
+type REJECTED = 2
+```
+
+### `inclusion-store` schema
+
+```typescript
+interface Inclusion {
```

Comment:
there is a comment there

```
// TODO: inclusion proof?
```

I think it will need to be a CID and we will store in a bucket as data structure can be quite big. But need to see it in real life